### ISSUE TITLE
"command not found" under npm2

### ISSUE THREAD

**Issue by jasonkarns** [zkat/npx#109]
I happened to be on a package that is still configured to use node4 (using nodenv). This issue doesn't manifest on npm3.

```
$ npx -c 'echo hi'
npx: command not found: echo hi
```

```
$ node --version
v4.4.7
$ npm --version
2.15.8
```

This isn't a big issue for me, as most of my other modules are using npm3 or later. However, as npm2 is still shipped with node 4 (LTS until April 2018) I would hope that it is still supported.

**Comment by zkat** [comment]
@jasonkarns this is... particularly weird because standalone npx bundles its own version of npm when installed standalone. Are you somehow using the version bundled with npm, then switching to npm2?

**Comment by jasonkarns** [comment]
**TL;DR: I've isolated the issue to versions of node prior to 4.8**

---
> Are you somehow using the version bundled with npm, then switching to npm2?

That's not really possible with nodenv. All global modules are installed in each node version's `/lib/node_modules` directory. So the only global node modules available are those that are installed in the currently-active node version.

For instance, asking nodenv which npx is currently running points to the one under node 4.4.7

```
$ nodenv which npx
/usr/local/var/nodenv/versions/4.4.7/bin/npx
$ npx -v
9.6.0
```

If I were to switch nodes to a version that doesn't have npx installed, it reports that npx is not currently available, but lists all the node versions that _do_ have that binary:

<details><pre><code>
$ nodenv shell 4.4.3 
$ node -v
v4.4.3
$ nodenv which npx
nodenv: npx: command not found

The `npx' command exists in these Node versions:
  4
  4.4
  4.4.7
  4.4.7-npm3
  4.8
  4.8.3
  6.9
  6.9.5
  8
  8.0
  8.0.0
  argon
</pre></code></details>

---
So I did a completely fresh install of node 4.4.7, leaving npm untouched so it's using the version that ships with node 4.4.7. Then I installed npx globally (which installs npx only for that node).

<details><pre><code>
$ nodenv uninstall 4.4.7
nodenv: remove /usr/local/var/nodenv/versions/4.4.7? [yN] y
$ nodenv install 4.4.7
Downloading node-v4.4.7-darwin-x64.tar.gz...
-> https://nodejs.org/dist/v4.4.7/node-v4.4.7-darwin-x64.tar.gz
Installing node-v4.4.7-darwin-x64...
Installed node-v4.4.7-darwin-x64 to /usr/local/var/nodenv/versions/4.4.7
$ npm i -g npx
/usr/local/var/nodenv/versions/4.4.7/bin/npx -> /usr/local/var/nodenv/versions/4.4.7/lib/node_modules/npx/index.js
npx@9.6.0 /usr/local/var/nodenv/versions/4.4.7/lib/node_modules/npx
$ node -v; npm -v; npx -v
v4.4.7
2.15.8
9.6.0
$ npx -c 'echo hi'
npx: command not found: echo hi
</code></pre></details>

---
I tried incrementally upgrading npm to each of the 2.x patch releases since 2.15.8, no change. (2.15 .9, .10, .11, .12)

I then tried incrementally upgrading node (using the stock npm with each node) and installing npx globally after each install. Each of node 4.5.0, 4.6.2, 4.7.3 all failed with same error. BUT! **Beginning with version 4.8.0, the npx command works without error**.

```
$ node -v; npm -v; npx -v
v4.8.0
2.15.11
9.6.0
$ npx -c 'echo hi'
hi
```

(All of the nodes mentioned above were installed fresh, so a fresh and clean lib/node_modules for each version.)

So it appears that the issue is limited to versions of node _prior_ to 4.8.0. And given that there are now 4 patch releases of 4.8, I'm assuming that there really isn't much value in attempting to resolve this for pre 4.8 nodes? If so, feel free to close this. Otherwise, I'll provide as much info as is helpful.

**Comment by zkat** [comment]
Since npx has now dropped support for node@4 altogether, I'm closing this.
